### PR TITLE
TF: XLA-trainable DeBERTa v2

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

As discussed in https://github.com/huggingface/transformers/issues/18476 and https://github.com/huggingface/transformers/issues/18239, there are two problems while training DeBERTa v2 with TensorFlow:
1. `TFDebertaV2StableDropout` doesn't work at training time (actually, its logic is only triggered at training time, so it doesn't work at all :D)
2. TF complains about unknown shapes in `take_along_axis` (forward and backward passes, when the batch dim is `None`)

This PR fixes both problems above :) 

Problem 1. is got a straightforward fix. The gradient propagation code didn't have the right gradient shapes -- this PR simplifies and fixes it by moving all functions inside the special dropout class (compare to the original PT implementation [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/deberta_v2/modeling_deberta_v2.py#L173) -- also notice how much more elegant TF's code is ;)).

Problem 2. is tricker. The exception gets fixed with the addition of a `shape_list`, but the code is super slow on TPU. This PR adds an if/else pair of branches, one that is efficient on TPU, the other on GPU :)

_____________________________________________________

⚠️ These exceptions were not caught because deberta v2 and v3 rely on special config options -- e.g. https://huggingface.co/microsoft/deberta-v3-base/blob/main/config.json#L14

How can we ensure we properly test these configurations?

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
Thanks for fixing!

**Comment by WissamAntoun** [comment]
@gante I think it's better to replace 

```python
flat_x = tf.reshape(x, (-1, x.shape[-1]))
flat_indices = tf.reshape(indices, (-1, indices.shape[-1]))
gathered = tf.gather(flat_x, flat_indices, batch_dims=1)
gathered = tf.reshape(gathered, shape_list(indices))
```
with 
```python
gathered = tf.gather(x,indices,batch_dims=2)
```
which gives the same numerical results and the same performance according to my tests

https://github.com/huggingface/transformers/issues/18239#issuecomment-1193126061

**Comment by gante** [comment]
@WissamAntoun thank you for pointing it out, I completely missed it in the original thread! 🙏 Will make the change

File: src/transformers/models/deberta_v2/modeling_tf_deberta_v2.py [hunk 1]
```diff
@@ -138,9 +115,26 @@ def __init__(self, drop_prob, **kwargs):
...
+        input = tf.cond(self.drop_prob > 0, lambda: tf.where(mask, 0.0, input) * scale, lambda: input)
+
+        def grad(upstream):
+            return tf.cond(scale > 1, lambda: tf.where(mask, 0.0, upstream) * scale, lambda: upstream)
+
+        return input, grad
+
     def call(self, inputs: tf.Tensor, training: tf.Tensor = False):
-        if training and self.drop_prob > 0:
-            return TFDebertaV2XDropout(inputs, self.drop_prob)
```
Comment (Rocketknight1) [inline src/transformers/models/deberta_v2/modeling_tf_deberta_v2.py:143]: Aaaa, was this instantiating a new class in each `call`?!

File: src/transformers/models/deberta_v2/modeling_tf_deberta_v2.py [hunk 2]
```diff
@@ -138,9 +115,26 @@ def __init__(self, drop_prob, **kwargs):
...
+        """
+        mask = tf.cast(
+            1 - tf.compat.v1.distributions.Bernoulli(probs=1 - self.drop_prob).sample(sample_shape=shape_list(input)),
+            tf.bool,
+        )
+        scale = tf.convert_to_tensor(1.0 / (1 - self.drop_prob), dtype=tf.float32)
+        input = tf.cond(self.drop_prob > 0, lambda: tf.where(mask, 0.0, input) * scale, lambda: input)
+
+        def grad(upstream):
+            return tf.cond(scale > 1, lambda: tf.where(mask, 0.0, upstream) * scale, lambda: upstream)
```
Comment (Rocketknight1) [inline src/transformers/models/deberta_v2/modeling_tf_deberta_v2.py:131]: Is there a reason to use `tf.cond` over standard Python `if` statements here? It feels harder to read, and `self.drop_prob` and `scale` shouldn't vary between calls!

**Review by Rocketknight1 — APPROVED** [review]
Looks great overall, thank you for finally fixing this!

File: src/transformers/models/deberta_v2/modeling_tf_deberta_v2.py (hunk 1 above)
Comment (gante) [reply src/transformers/models/deberta_v2/modeling_tf_deberta_v2.py:143]: not quite because `TFDebertaV2XDropout`, despite the camel case, was a function 😂

File: src/transformers/models/deberta_v2/modeling_tf_deberta_v2.py (hunk 2 above)
Comment (gante) [reply src/transformers/models/deberta_v2/modeling_tf_deberta_v2.py]: The original code converted `self.drop_prob` into a tensor, so all if/elses had to be `tf.cond`. 

But there is no good reason for `self.drop_prob` to be a tensor, so I've removed the conversion and replaced `tf.cond` by if/elses! Thanks for the suggestion 👍
